content/blog/2021/11/deep-dive-into-policy-logging.md

```diff
@@ -0,0 +1,221 @@
+file is to initialize the logger:
+
+```go
+var (
+	logWriter = kubewarden.KubewardenLogWriter{}
+	logger    = onelog.New(
+		&logWriter,
+		onelog.ALL, // shortcut for onelog.DEBUG|onelog.INFO|onelog.WARN|onelog.ERROR|onelog.FATAL
+	)
+)
+```
+
+We are then able to use `onelog` API in order to produce log messages. We could, for example,
+perform structured logging with debugging level:
+
+```go
+logger.DebugWithFields("validating object", func(e onelog.Entry) {
+    e.String("name", gjson.GetBytes(payload, "request.object.metadata.name").String())
+	e.String("namespace", gjson.GetBytes(payload, "request.object.metadata.namespace").String())
+})
+```
+
+Or, with info level:
+
+```go
+logger.InfoWithFields("validating object", func(e onelog.Entry) {
+    e.String("name", gjson.GetBytes(payload, "request.object.metadata.name").String())
+	e.String("namespace", gjson.GetBytes(payload, "request.object.metadata.namespace").String())
+})
+```
+
+What happens under the covers is that our Go SDK sends [every log event to the `kubewarden` host
+through
+waPC](https://github.com/kubewarden/policy-sdk-go/blob/95c4b93adf6830862b85f953bc09d31de7490872/log_writer_wasi.go#L15).
+
+### Rust
+
+Let's use the [Rust policy template](https://github.com/kubewarden/policy-rust-template) as our
+guide.
+
+Our Rust SDK implements an integration with the `slog` crate. This crate exposes the concept of
+drains, so we have to [define a global drain that we will use throughout our policy
+code](https://github.com/kubewarden/policy-rust-template/blob/fd46e75b3794dbacbf499da6632d5294bdcea28c/src/lib.rs#L16-L21):
+
+```rust
+use kubewarden::logging;
+use slog::{o, Logger};
+lazy_static! {
+    static ref LOG_DRAIN: Logger = Logger::root(
+        logging::KubewardenDrain::new(),
+        o!("some-key" => "some-value") // This key value will be shared by all logging events that use
+                                       // this logger
+    );
+}
+```
+
+Then, we can use the macros provided by `slog` to log on different levels:
+
+```rust
+use slog::{crit, debug, error, info, trace, warn};
+```
+
+Let's log an info level message:
+
+```rust
+info!(
+    LOG_DRAIN,
+    "rejecting resource";
+    "resource_name" => &resource_name
+);
+```
+
+As happens with the Go SDK implementation, our Rust implementation of the `slog` drain sends [this
+logging events to the host by using
+waPC](https://github.com/kubewarden/policy-sdk-rust/blob/642d13b07053be6455c0b5f49df0e87ff6f022c5/src/logging/drain.rs#L57-L60).
+
+You can read more about slog [here](https://github.com/slog-rs/slog).
+
+### Swift
+
+We will be looking at the [Swift policy
+template](https://github.com/kubewarden/swift-policy-template) for this example.
+
+As happens with Go and Rust's SDK's, the Swift SDK is instrumented to use Swift's `LogHandler`, so
+our policy only has to initialize it. In our `Sources/Policy/main.swift` file:
```

Comment:
We're using the popular [swift-log](https://github.com/apple/swift-log). Can you write that and drop a link to the GitHub repo?

Reply:
The link was some lines after, I can add another link here.